### ISSUE TITLE
Unable to recognize external USB device

### ISSUE THREAD

**Issue by lingyinsam** [fbelavenuto/arpl#329]
I come from China, my English is very poor, the following words, I was written in Chinese, with the translation tool translated into English, I hope you can understand.
1. My equipment: G4560+B365SD4, a PCIE to 10 SATA HUB, there are 3 SATA interfaces on the motherboard. There are 2+1USB3.0 and 2+2USB2.0.
2. My problem: DSM recognizes my device as having 26 SATA. When I plug in an external storage device to any USB port, it will be recognized as internal storage. DSM prompts me to create a new storage pool to use. This prevents me from transferring files using external storage devices. How to solve it?
3. Finally, how to adjust SataPortMap, DiskldxMap and SasldxMap?

**Comment by fbelavenuto** [comment]
Please go to menu, choose "cmdline menu" and choose "Show SATA(s) # ports and drives". Post screen photo here

**Comment by lingyinsam** [comment]
![QQ截图20221119095413](https://user-images.githubusercontent.com/69802837/202828799-f36ccea4-a00a-44cf-b174-c737921bea1d.png)
The first SATA controller is the motherboard, it shows that there are 6 SATA ports, but my motherboard only has 3 SATA interfaces
The second is an external SATA HUB, which has only 10 SATA interfaces. When booting and loading the HUB, the 1-6 interface is shown as 7-12 in the figure. I only insert 1 hard disk into the 7-10 interface, which is displayed as 1F, corresponding to 38 in the figure.

**Comment by lingyinsam** [comment]
![QQ截图20221119095413](https://user-images.githubusercontent.com/69802837/202829117-28e16226-4c92-48a3-9b21-eb88816f15dd.png)
The first SATA controller belongs to the motherboard, which shows that there are 6 SATA ports, but my motherboard only has 3 SATA interfaces.
The second is an external SATA HUB, which has only 10 SATA interfaces. When booting and loading the HUB, the 1-6 interface is shown as 7-12 in the figure. I only insert 1 hard disk into the 7-10 interface, which is displayed as 1F, corresponding to 38 in the figure.

**Comment by lingyinsam** [comment]
![QQ截图20221119095413](https://user-images.githubusercontent.com/69802837/202829329-29a6220e-0c80-4b54-ad5a-127758c50ea0.png)
The first SATA controller is the motherboard, which shows 6 SATAs, but my motherboard only has 3 SATA interfaces and 1 M.2 interface (NVME/M.2 SATA).
Interface No. 4 in the picture is my M.2 SATA boot disk.
The second SATA controller is an external SATA HUB, and this HUB only has 10 SATA interfaces.
When the HUB is loaded at startup, 1-6 interfaces correspond to 7-12 in the figure.
In the 7-10 interface of the HUB, I only insert 1 hard disk, which is displayed as 1F, corresponding to 38 in the figure.

**Comment by fbelavenuto** [comment]
Ok, this ASMedia have 32 ports internally, is much!! Physically have only 10 ports?

**Comment by lingyinsam** [comment]
Yes, but it doesn't quite get the ports sorted. 1-6 total 6 port sorting is 01,02,03,04,05,06,7-10 sorting should be 1F... like hexadecimal.

**Comment by lingyinsam** [comment]
![1669030411916](https://user-images.githubusercontent.com/69802837/203041038-3b9a57bf-d5b4-4da6-92b0-e8d23c6b1367.jpg)
like this.

**Comment by lingyinsam** [comment]
This is the prompt after I insert the U disk.
![QQ截图20221121194219](https://user-images.githubusercontent.com/69802837/203043205-521fc9f3-c775-4438-a23c-d4a67afcbac1.png)
![QQ截图20221121194324](https://user-images.githubusercontent.com/69802837/203043218-8ff992fe-5e01-48ed-a29a-485050eabee4.png)

**Comment by fbelavenuto** [comment]
This SATA card created unnecessary internal ports, "SataPortMap" and "DiskIdxMap" variables will not work correctly.

Please map the tree missing ports so we know what the number of each one will be. Four port is number 38 (0x1F hexa relative)

**Comment by lingyinsam** [comment]
Do you mean that I need to determine the ports corresponding to 7-10 of this SATA HUB?

**Comment by fbelavenuto** [comment]
Yes!! One of them you already determined.

**Comment by lingyinsam** [comment]
I'm going to check now.

**Comment by lingyinsam** [comment]
I made sure that the 7-10 of this HUB corresponds to 35-38.
![35](https://user-images.githubusercontent.com/69802837/203061229-e3db09e2-abc4-4643-b83a-0d70a838d789.png)
![36](https://user-images.githubusercontent.com/69802837/203061234-c7991ea2-9d5e-4637-8c5b-b6211369b864.png)
![37](https://user-images.githubusercontent.com/69802837/203061243-53390dbd-b42a-4abd-8c59-fe2fce319193.png)

**Comment by lingyinsam** [comment]
So what do I need to do now to get it working?

**Comment by fbelavenuto** [comment]
Do a test, add the SataPortMap cmdline, set the value to 46
Look here how do it:
https://github.com/fbelavenuto/arpl/issues/103
This will remove ports 13-38 (absolute), or relative ports 7-32, from extra SATA board.

**Comment by lingyinsam** [comment]
I gave it a try. Now booting up at 7-10 and inserting the hard disk DSM does not recognize it, and there is no response.

**Comment by fbelavenuto** [comment]
Ok, try with external storage into USB ports

**Comment by lingyinsam** [comment]
DSM can recognize USB external storage now.
![45](https://user-images.githubusercontent.com/69802837/203068188-935407d6-3213-4c49-8172-cdb028cc3484.png)

**Comment by lingyinsam** [comment]
![46](https://user-images.githubusercontent.com/69802837/203068300-729db7be-23ab-4829-8881-c3cd210f8a4e.png)

**Comment by fbelavenuto** [comment]
Please what model of DSM are you using?

**Comment by lingyinsam** [comment]
DS3617xs

**Comment by lingyinsam** [comment]
The DSM version is DSM 7.1.1-42962 Update 2.

**Comment by lingyinsam** [comment]
I remember before the 7-10 when the hard drive was not plugged in, the USB external storage was normal.

**Comment by fbelavenuto** [comment]
Ok, now remove the SataPortMap and add the variable "usbportcfg" with value "0x3C000000000"

**Comment by lingyinsam** [comment]
![222222222222](https://user-images.githubusercontent.com/69802837/203070919-d4b9cff1-1c48-48cd-b5f3-491f16ca5844.png)
![333333333333](https://user-images.githubusercontent.com/69802837/203070924-b8b033fb-daff-4c28-9710-25832108691f.png)
How to delete? Enter directly?

**Comment by lingyinsam** [comment]
![2535352](https://user-images.githubusercontent.com/69802837/203071557-429ffd55-0482-4588-a494-eb6af8f84053.png)
I try to add usbportcfg, then remove SataPortMap, then as shown.

**Comment by fbelavenuto** [comment]
Select the two entries (first is wrong, last is correct)

**Comment by lingyinsam** [comment]
I don't know how to select and delete.

**Comment by lingyinsam** [comment]
![QQ截图20221121215746](https://user-images.githubusercontent.com/69802837/203073190-94397141-d436-4632-aafb-059c72bab270.png)
How to select the brackets in front of this?

**Comment by lingyinsam** [comment]
![QQ截图20221121220504](https://user-images.githubusercontent.com/69802837/203074729-c261d942-90a6-44d5-8de5-4bfa12d1f716.png)
I learned. Sorry, I didn't know this page could still be selected with the mouse.

**Comment by lingyinsam** [comment]
I gave it a try. DSM is still recognized as a local disk even if the USB external storage device is inserted.
![QQ截图20221121220810](https://user-images.githubusercontent.com/69802837/203075516-686861a7-e61b-41d5-bb9c-daef51de5395.png)

**Comment by fbelavenuto** [comment]
Ok so I need to study more to see how to solve your problem

**Comment by fbelavenuto** [comment]
> If I change the SataPortMap to 49, will it work?

If your define the SataPortMap with this value, will happens:

![image](https://user-images.githubusercontent.com/11461290/203320416-57e2f7b6-a820-43c1-a9a8-4b0b45680127.png)

In first line only the first four ports will be recognized and in second line only the first 9 ports will be recognized.

The problem is your SATA card (second line), it has too many unused ports in between!!

**Comment by fbelavenuto** [comment]
Closing, original question are answered
